File: src/orchestrator/confidence.ts
export interface ConfidenceInputs {
    modelConfidence?: number;
    verifierConfidence?: number;
    embeddingScore?: number;
}

/**
 * Combine multiple confidence signals into a single score.
 * All inputs are expected to be in the range 0.0 – 1.0.
 */
export function combineConfidence(
    inputs: ConfidenceInputs
): number {
    const weights = {
        model: 0.3,
        verifier: 0.5,
        embedding: 0.2,
    };

    let score = 0;
    let totalWeight = 0;

    if (typeof inputs.modelConfidence === 'number') {
        score += inputs.modelConfidence * weights.model;
        totalWeight += weights.model;
    }

    if (typeof inputs.verifierConfidence === 'number') {
        score += inputs.verifierConfidence * weights.verifier;
        totalWeight += weights.verifier;
    }

    if (typeof inputs.embeddingScore === 'number') {
        score += inputs.embeddingScore * weights.embedding;
        totalWeight += weights.embedding;
    }

    if (totalWeight === 0) {
        return 0;
    }

    return clamp(score / totalWeight, 0, 1);
}

/**
 * Decide whether confidence is sufficient to return an answer.
 */
export function isConfidenceAcceptable(
    confidence: number,
    threshold: number = 0.75
): boolean {
    return confidence >= threshold;
}

/**
 * Decide whether escalation is required.
 */
export function shouldEscalate(
    confidence: number,
    minConfidence: number = 0.5
): boolean {
    return confidence < minConfidence;
}

/**
 * Clamp values safely.
 */
function clamp(
    value: number,
    min: number,
    max: number
): number {
    return Math.max(min, Math.min(max, value));
}

File: rust/rlm/src/orchestrator/confidence.rs
#[derive(Debug, Clone, Default)]
pub struct ConfidenceInputs {
    pub model_confidence: Option<f64>,
    pub verifier_confidence: Option<f64>,
    pub embedding_score: Option<f64>,
}

pub fn combine_confidence(inputs: ConfidenceInputs) -> f64 {
    let weights = (0.3_f64, 0.5_f64, 0.2_f64);

    let mut score = 0.0;
    let mut total_weight = 0.0;

    if let Some(c) = inputs.model_confidence {
        score += c * weights.0;
        total_weight += weights.0;
    }

    if let Some(c) = inputs.verifier_confidence {
        score += c * weights.1;
        total_weight += weights.1;
    }

    if let Some(c) = inputs.embedding_score {
        score += c * weights.2;
        total_weight += weights.2;
    }

    if total_weight == 0.0 {
        return 0.0;
    }

    clamp(score / total_weight, 0.0, 1.0)
}

pub fn is_confidence_acceptable(confidence: f64, threshold: f64) -> bool {
    confidence >= threshold
}

pub fn should_escalate(confidence: f64, min_confidence: f64) -> bool {
    confidence < min_confidence
}

fn clamp(value: f64, min: f64, max: f64) -> f64 {
    if value < min {
        min
    } else if value > max {
        max
    } else {
        value
    }
}
